File: ueueo-framework/ueueo-background-jobs/src/main/java/com/ueueo/backgroundjobs/DefaultBackgroundJobManager.cs
using System;
using System.Threading.Tasks;
using Volo.Abp.DependencyInjection;
using Volo.Abp.Guids;
using Volo.Abp.Timing;

namespace Volo.Abp.BackgroundJobs;

/**
 * Default implementation of <see cref="IBackgroundJobManager"/>.
*/
[Dependency(ReplaceServices = true)]
public class DefaultBackgroundJobManager : IBackgroundJobManager, ITransientDependency
{
    protected IClock Clock;//  { get; }
    protected IBackgroundJobSerializer Serializer;//  { get; }
    protected IGuidGenerator GuidGenerator;//  { get; }
    protected IBackgroundJobStore Store;//  { get; }

    public DefaultBackgroundJobManager(
        IClock clock,
        IBackgroundJobSerializer serializer,
        IBackgroundJobStore store,
        IGuidGenerator guidGenerator)
    {
        Clock = clock;
        Serializer = serializer;
        GuidGenerator = guidGenerator;
        Store = store;
    }

    public    Task<String> EnqueueAsync<TArgs>(TArgs args, BackgroundJobPriority priority = BackgroundJobPriority.Normal, TimeSpan? delay = null)
    {
        var jobName = BackgroundJobNameAttribute.GetName<TArgs>();
        var jobId = EnqueueAsync(jobName, args, priority, delay);
        return jobId.ToString();
    }

    protected    Task<Guid> EnqueueAsync(String jobName, Object args, BackgroundJobPriority priority = BackgroundJobPriority.Normal, TimeSpan? delay = null)
    {
        var jobInfo = new BackgroundJobInfo
        {
            Id = GuidGenerator.Create(),
            JobName = jobName,
            JobArgs = Serializer.Serialize(args),
            Priority = priority,
            CreationTime = Clock.Now,
            NextTryTime = Clock.Now
        };

        if (delay.HasValue)
        {
            jobInfo.NextTryTime = Clock.Now.Add(delay.Value);
        }

        Store.InsertAsync(jobInfo);

        return jobInfo.Id;
    }
}

File: ueueo-frameworks/ueueo-background-jobs/src/main/java/com/ueueo/backgroundjobs/DefaultBackgroundJobManager.java
package com.ueueo.backgroundjobs;

import com.ueueo.ID;
import com.ueueo.guids.IGuidGenerator;

import java.util.Calendar;
import java.util.Date;
import java.util.concurrent.TimeUnit;

/**
 * Default implementation of <see cref="IBackgroundJobManager"/>.
 */
public class DefaultBackgroundJobManager implements IBackgroundJobManager {
    protected IBackgroundJobSerializer backgroundJobSerializer;
    protected IGuidGenerator guidGenerator;
    protected IBackgroundJobStore backgroundJobStore;

    public DefaultBackgroundJobManager(
            IBackgroundJobSerializer backgroundJobSerializer,
            IBackgroundJobStore backgroundJobStore,
            IGuidGenerator guidGenerator) {
        this.backgroundJobSerializer = backgroundJobSerializer;
        this.guidGenerator = guidGenerator;
        this.backgroundJobStore = backgroundJobStore;
    }

    @Override
    public ID enqueue(String jobName, Object args, BackgroundJobPriority priority, Integer delay, TimeUnit delayTimeUnit) {

        BackgroundJobInfo jobInfo = new BackgroundJobInfo();
        jobInfo.setId(guidGenerator.create());
        jobInfo.setJobName(jobName);
        jobInfo.setJobArgs(backgroundJobSerializer.serialize(args));
        jobInfo.setCreationTime(new Date());
        jobInfo.setNextTryTime(new Date());
        jobInfo.setPriority(priority);

        if (delay != null && delay > 0) {
            Calendar calendar = Calendar.getInstance();
            calendar.add(Calendar.MILLISECOND, (int) delayTimeUnit.toMillis(delay));
            jobInfo.setNextTryTime(calendar.getTime());
        }

        backgroundJobStore.insert(jobInfo);

        return jobInfo.getId();
    }

}
